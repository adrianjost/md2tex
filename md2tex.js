#!/usr/bin/env node

"use-strict";

// Copyright (C) 2019 Adrian Jost
// This code is licensed under MIT license (see https://tldrlegal.com/license/mit-license for details)

import dd from "dedent-tabs";
import logLib from "./logger";

let logger = logLib();

const escapeSpecialChars = text => {
	text = text.replace(/([\\])/g, "\\textbackslash ");
	text = text.replace(/([\~])/g, "\\textasciitilde ");
	text = text.replace(/([\^])/g, "\\textasciicircum ");
	const prefix = char => `\\${char}`;
	text = text.replace(/([\&\%\$\#\_\{\}])/g, prefix);
	return text;
};
const removeSpecialChars = text =>
	text.replace(/([\&\%\$\#\_\{\}\~\^\\])/g, "");

const mapHeadlines = (line, index) => {
	let content = line.replace(/^[#]+\W/, "");
	const text = mapInlineTextStyles(content);
	const identifier =
		`${index}-` + removeSpecialChars(content.toLowerCase()).replace(/\W/g, "-");
	const level = line.match(/^[#]+/g)[0].length;

	switch (level) {
		case 1:
			return `%************************************************\n\\chapter{${text}}\n\\label{ch:${identifier}}\n%************************************************`;
		case 2:
			return `\\hypertarget{${identifier}}{\n\\section{${text}}\\label{${identifier}}}`;
		case 3:
			return `\\hypertarget{${identifier}}{\n\\subsection{${text}}\\label{${identifier}}}`;
		case 4:
			return `\\hypertarget{${identifier}}{\n\\subsubsection{${text}}\\label{${identifier}}}`;
		case 5:
			return `\\hypertarget{${identifier}}{\n\\paragraph{${text}}\\label{${identifier}}}`;
		case 6:
			return `\\hypertarget{${identifier}}{\n\\subparagraph{${text}}\\label{${identifier}}}`;
		default:
			logger.error(`unsupported heading ${index}: "${line}"`);
			return line;
	}
};

const mapCodeblocks = (line, index) => {
	if (!line.startsWith("```")) {
		return line;
	}
	if (line.trim() === "```") {
		return `	\\end{minted}
	\\caption{Code Snipped ${index}}
	\\label{lst:code-snipped-${index}}
\\end{listing}`;
	}
	const language = line
		.slice(3)
		.trim()
		.toLowerCase();

	return `\\begin{listing}[H]
	\\begin{minted}{${language}}`;
};

const mapInlineCodeblocks = line => {
	const inlineCode = function() {
		const args = Array.from(arguments);
		return `\\colorbox{gray-light!}{\\texttt{${escapeSpecialChars(args[1])}}}`;
	};
	return line.replace(/\`([^\`]*)\`/g, inlineCode);
};

const mapTextStyle = line => {
	const replaceWithUnescapedMatch = newString =>
		function(match) {
			let out = newString;
			Array.from(arguments).forEach((a, i) => {
				// mimic .replace api
				if (typeof a !== "string") {
					return;
				}
				// unescape
				const unescaped = a.replace(/\\([\*\_\\])/g, "$1");
				out = out.replace(`$${i}`, unescaped);
			});
			return out;
		};

	// 1. escape all _ and * inside urls
	// https://regex101.com/r/TJ2xDV/3
	line = line.replace(/(?<=[^\!]\[.*\]\(.*)([\\\_\*])(?=.*\))/g, "\\$1");

	// 2. apply bold when not escaped. then unescape the text
	line = line
		// check regex for single character below for details
		.replace(
			/(?<=(?:^|[^\\])(?:\\\\)*)(?<!\!\[.*\].*)(\*\*|\_\_)((?:[^\\]|\\[^\\]|\\\\)*?)\1/g,
			replaceWithUnescapedMatch("\\textbf{$2}")
		);
	// 3. apply italic when not escaped. then unescape the text
	line = line
		// https://regex101.com/r/jocMku/2
		.replace(
			/(?<=(?:^|[^\\])(?:\\\\)*)(?<!\!\[.*\].*)(\*|\_)((?:[^\\]|\\[^\\]|\\\\)*?)\1/g,
			replaceWithUnescapedMatch("\\textit{$2}")
		);
	// 4. unescape all text inside links
			line = line.replace(/(?<=[^\!]\[.*\]\(.*)\\([\_\*])(?=.*\))/g, "$1")

	return line;
};

const mapImages = line => {
	const replaceImage = function(match) {
		const args = Array.from(arguments);
		return `\n\\begin{figure}[H]
	\\centering
	\\includegraphics[width=\\textwidth]{${args[2]}}
	\\caption[${removeSpecialChars(args[1])}]{${removeSpecialChars(args[1])}}
	\\label{fig:${removeSpecialChars(args[1])}}
\\end{figure}\n`;
	};
	return line.replace(/\!\[([^\]]*)\]\(([^\)]+)\)/g, replaceImage);
};

const mapFootnotes = line => {
	const footnote = function(){
		const args = Array.from(arguments);
		return `${escapeSpecialChars(args[1])}\\footnote{${escapeSpecialChars(args[2])}}`
	}
	return line.replace(/(?<!\!)\[([^\]]*)\]\(([^\)]+)\)/g, footnote);
};

const mapInlineTextStyles = (line, index) => {
	line = mapTextStyle(line);
	line = mapInlineCodeblocks(line, index);
	line = mapFootnotes(line);
	return line;
};

export const convert = (content, { loglevel = "warn" } = {}) => {
	logger = logLib(loglevel);
	content = content + "\n".repeat(2);
	let inRawLatexBlock = false;
	let inCodeBlock = false;
	let openEnums = [];
	const enums = [
		[/^(\s)*[1-9]+\.(\s)/, "enumerate"],
		[/^(\s)*\-(\s)/, "itemize"]
	];
	let lines = content
		.split("\n")
		.map(a => a.trimRight().replace(/(\r\n|\n|\r)$/gm, ""))
	lines = lines
		.map((a, i) => {
			// latex block
			if (a.toLowerCase().startsWith("```latex")) {
				inRawLatexBlock = true;
				return "";
			} else if (inRawLatexBlock && a.startsWith("```")) {
				inRawLatexBlock = false;
				return "";
			} else if (inRawLatexBlock) {
				return a;
			}

			// code block styles
			if (a.startsWith("```")) {
				inCodeBlock = !inCodeBlock;
			}
			a = mapCodeblocks(a);
			if (inCodeBlock) {
				return a;
			}

			if (a.startsWith("#")) {
				return mapHeadlines(a, i);
			}
			// text styles
			a = mapInlineTextStyles(a, i);
			a = mapImages(a);

			// Lists
			const activeEnum = enums.find(([regex]) => {
				return a.match(regex);
			});

			if (activeEnum) {
				const [regex, type] = activeEnum;

				const leadingSpaces = a.search(/\S/);
				a = a.replace(
					regex,
					(match, m1, m2) =>
						`${"\t".repeat(leadingSpaces + 1)}\\item${m2}`
				);

				// more indentation than before
				if (leadingSpaces + 1 > openEnums.length) {
					a = `${"\t".repeat(leadingSpaces)}\\begin{${type}}\n${a}`;
					openEnums.push(type);
					// less indentation than before
				} else if (leadingSpaces < openEnums.length) {
					const diff = openEnums.length - leadingSpaces - 1;
					let closingTags = [];
					for (let index = 0; index < diff; index++) {
						const endtype = openEnums.pop();
						closingTags.push(`${"\t".repeat(diff - index)}\\end{${endtype}}`);
					}
					a = `${closingTags.join("\n")}\n${a}`;
				}
			} else if (openEnums.length) {
				a = `${openEnums
					.map(
						(type, index) =>
							`${"\t".repeat(openEnums.length - index - 1)}\\end{${type}}`
					)
					.join("\n")}\n${a} ${(lines[Math.min(i + 1, lines.length - 1)].replace(/\s/g, "").length && a.replace(/\s/g, "").length) ? "\\\\" : ""}`;
				openEnums = [];
			}else if(lines[Math.min(i + 1, lines.length - 1)].replace(/\s/g, "").length && a.replace(/\s/g, "").length ){
				a = `${a} ${"\\\\"}`;
			}

			// Return Result
			return a;
		});
	const output = lines.slice(0, -1).join("\n");
	logger.log("---\n"+output+"\n---");
	return output;
};

export default convert;
