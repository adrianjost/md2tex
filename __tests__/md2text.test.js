import { convert as converter } from "../md2tex";
import dd from "dedent-tabs";

const convert = (content, config = { loglevel: "log"}) => {
	return converter(content, config);
}

describe("md2tex", () => {
	it("headlines", () => {
		expect(
			convert(dd`
				# H1
				## H2
				### H3
				#### H4
				##### H5
				###### H6
			`)
		).toMatchSnapshot();
	});
	/*
	it("headlines with escaped characters", () => {
		expect(
			convert(dd`
				# escapes
				## blocks
				### ~ tilde ~
				### ^I don't know
				### \\ backslash \\
				## & other &
				### % procent %
				### $ money $ dollar
				### # hashtag #
				### _ underscore _
				###### ({brackets})
			`)
		).toMatchSnapshot();
	});
	*/
	it("headlines with styles and code", () => {
		expect(
			convert(dd`
				# *italic* _underscore_
				## \`Code\` headline
			`)
		).toMatchSnapshot();
	});
	it("inline styles", () => {
		expect(
			convert(dd`
				*italic* normal _italic_
				**bold** normal __bold__
			`)
		).toMatchSnapshot();
	});
	it("inline styles with links and images", () => {
		expect(
			convert(dd`
				*italic* [link](without_any_italic_styles) _italic_
				**bold** ![images](shouldn**t**be__bold__eighter).
			`)
		).toMatchSnapshot();
	});
	it("footnotes", () => {
		expect(convert(dd`some [footnote](footnote text)`)).toMatchSnapshot();
	});
	it("images", () => {
		expect(convert(dd`some ![image](url)`)).toMatchSnapshot();
	});
	it("ordered list", () => {
		expect(
			convert(dd`
		1. first entry
			1. first indentation
				1. even more indentation
		2. second entry
			1. indented
				1. and done
		`)
		).toMatchSnapshot();
	});
	it("unordered list", () => {
		expect(
			convert(dd`
		- first entry
			- first indentation
				- even more indentation
		- second entry
			- indented
				- and done
		`)
		).toMatchSnapshot();
	});
	it("mixed list", () => {
		expect(
			convert(dd`
		1. first entry
			- first indentation
				- even more indentation
			- second subentry
		2. second entry
			1. indented
				- unorderd sublist
		`)
		).toMatchSnapshot();
	});
	it("code blocks", () => {
		expect(
			convert(dd`
			some text before
			\`\`\`js
			console.log("Hi");
			\`\`\`
			and some behind
		`)
		).toMatchSnapshot();
	});
	it("raw latex blocks", () => {
		expect(
			convert(dd`
			some text before
			\`\`\`latex
			\\textbackslash backslash
			\`\`\`
			and some behind
		`)
		).toMatchSnapshot();
	});
	it("log error for unsupported heading levels", () => {
		expect(
			convert(dd`
			######## H8 is unsupported, errors are expected
		`)
		).toMatchSnapshot();
	});
});
