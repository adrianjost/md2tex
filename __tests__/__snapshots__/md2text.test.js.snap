// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md2tex code blocks 1`] = `
"some text before
\\\\begin{listing}[H]
	\\\\begin{minted}{js}
console.log(\\"Hi\\");
	\\\\end{minted}
	\\\\caption{Code SnippeCode Snipped undefined}
	\\\\label{lst:code-snipped-undefined}
\\\\end{listing}
and some behind
"
`;

exports[`md2tex footnotes 1`] = `
"some footnote
\\\\footnote{footnote text}

"
`;

exports[`md2tex headlines 1`] = `
"%************************************************
\\\\chapter{H1}
\\\\label{ch:0-h1}
%************************************************
\\\\hypertarget{1-h2}{
\\\\section{H2}\\\\label{1-h2}}
\\\\hypertarget{2-h3}{
\\\\subsection{H3}\\\\label{2-h3}}
\\\\hypertarget{3-h4}{
\\\\subsubsection{H4}\\\\label{3-h4}}
\\\\hypertarget{4-h5}{
\\\\paragraph{H5}\\\\label{4-h5}}
\\\\hypertarget{5-h6}{
\\\\subparagraph{H6}\\\\label{5-h6}}
"
`;

exports[`md2tex headlines with styles and code 1`] = `
"%************************************************
\\\\chapter{\\\\textit{italic} \\\\textit{underscore}}
\\\\label{ch:0--italic--underscore}
%************************************************
\\\\hypertarget{1--code--headline}{
\\\\section{\\\\colorbox{gray-light!}{\\\\texttt{Code}} headline}\\\\label{1--code--headline}}
"
`;

exports[`md2tex images 1`] = `
"some 
\\\\begin{figure}[H]
	\\\\centering
	\\\\includegraphics[width=\\\\textwidth]{url}
	\\\\caption[image]{image}
	\\\\label{fig:image}
\\\\end{figure}

"
`;

exports[`md2tex inline styles 1`] = `
"\\\\textit{italic} normal \\\\textit{italic}
\\\\textbf{bold} normal \\\\textbf{bold}
"
`;

exports[`md2tex inline styles with links and images 1`] = `
"\\\\textit{italic} link
\\\\footnote{without\\\\_any\\\\_italic\\\\_styles}
 \\\\textit{italic}
\\\\textbf{bold} 
\\\\begin{figure}[H]
	\\\\centering
	\\\\includegraphics[width=\\\\textwidth]{shouldn**t**be__bold__eighter}
	\\\\caption[images]{images}
	\\\\label{fig:images}
\\\\end{figure}
.
"
`;

exports[`md2tex log error for unsupported heading levels 1`] = `
"######## H8 is unsupported, errors are expected
"
`;

exports[`md2tex mixed list 1`] = `
"\\\\begin{enumerate}
	\\\\item first entry
	\\\\begin{itemize}
		\\\\item first indentation
		\\\\begin{itemize}
			\\\\item even more indentation
	\\\\end{itemize}
		\\\\item second subentry
	\\\\end{itemize}
	\\\\item second entry
	\\\\begin{enumerate}
		\\\\item indented
		\\\\begin{itemize}
			\\\\item unorderd sublist
		\\\\end{enumerate}
	\\\\end{enumerate}
\\\\end{itemize}
"
`;

exports[`md2tex ordered list 1`] = `
"\\\\begin{enumerate}
	\\\\item first entry
	\\\\begin{enumerate}
		\\\\item first indentation
		\\\\begin{enumerate}
			\\\\item even more indentation
		\\\\end{enumerate}
	\\\\end{enumerate}
	\\\\item second entry
	\\\\begin{enumerate}
		\\\\item indented
		\\\\begin{enumerate}
			\\\\item and done
		\\\\end{enumerate}
	\\\\end{enumerate}
\\\\end{enumerate}
"
`;

exports[`md2tex raw latex blocks 1`] = `
"some text before

\\\\\\\\textbackslash backslash

and some behind
"
`;

exports[`md2tex unordered list 1`] = `
"\\\\begin{itemize}
	\\\\item first entry
	\\\\begin{itemize}
		\\\\item first indentation
		\\\\begin{itemize}
			\\\\item even more indentation
		\\\\end{itemize}
	\\\\end{itemize}
	\\\\item second entry
	\\\\begin{itemize}
		\\\\item indented
		\\\\begin{itemize}
			\\\\item and done
		\\\\end{itemize}
	\\\\end{itemize}
\\\\end{itemize}
"
`;
